libsnark/gadgetlib1/gadgets/hashes/anemoi/tests/test_anemoi_gadget.cpp

```diff
@@ -224,18 +226,92 @@ void test_flystel_prime_field_gadget()
+    std::vector<FieldT> D;
+
+    pb_variable_array<FieldT> X_left;
+    pb_variable_array<FieldT> X_right;
+    pb_variable_array<FieldT> Y_left;
+    pb_variable_array<FieldT> Y_right;
+
+    X_left.allocate(pb, NumStateColumns_L, "left inputs");
+    X_right.allocate(pb, NumStateColumns_L, "right inputs");
+
+    Y_left.allocate(pb, NumStateColumns_L, "left outputs");
+    Y_right.allocate(pb, NumStateColumns_L, "right outputs");
+
+    // WARNINIG! test with zero constants. TODO add the original
+    // constants of Anemoi
+    for (size_t i = 0; i < NumStateColumns_L; i++) {
+        C.push_back(FieldT(0));
+        D.push_back(FieldT(0));
+    }
+
+    anemoi_permutation_round_prime_field_gadget<
+        ppT,
+        NumStateColumns_L,
+        parameters>
+        d(pb, C, D, X_left, X_right, Y_left, Y_right, "anemoi permutation");
+
+    // generate constraints
+    d.generate_r1cs_constraints();
+
+    // Input values: X_left = 0,1,2...L-1 ; X_right = L, L+1, 2L-1
+    for (size_t i = 0; i < NumStateColumns_L; i++) {
+        pb.val(X_left[i]) = FieldT(i);
+        pb.val(X_right[i]) = FieldT(NumStateColumns_L + i);
+    }
+
+    // generate witness for the given input
+    d.generate_r1cs_witness();
+
+    std::vector<FieldT> Y_expect =
+        anemoi_expected_output_one_round(NumStateColumns_L);
+
+    for (size_t i = 0; i < NumStateColumns_L; i++) {
+        ASSERT_EQ(Y_expect[i], pb.val(Y_left[i]));
+        ASSERT_EQ(Y_expect[NumStateColumns_L + i], pb.val(Y_right[i]));
+    }
+
+    ASSERT_TRUE(pb.is_satisfied());
+    test_pb_verify_circuit<ppT>(pb);
+
+    libff::print_time(
+        "anemoi_permutation_round_prime_field_gadget tests successful");
+}
+
 template<typename ppT> void test_for_curve()
 {
     // Execute all tests for the given curve.
 
     ppT::init_public_params();
-    using parameters = debug_parameters_bls12_381;
-
-    test_flystel_Q_gamma_prime_field_gadget<ppT, parameters>();
-    test_flystel_Q_gamma_binary_field_gadget<ppT, parameters>();
+    // Use debug parameters with small values for the small gadgets
+    using parameters_debug = parameters_debug_bls12_381;
+    test_flystel_Q_gamma_prime_field_gadget<ppT, parameters_debug>();
+    test_flystel_Q_gamma_binary_field_gadget<ppT, parameters_debug>();
     test_flystel_E_power_five_gadget<ppT>();
-    test_flystel_E_root_five_gadget<ppT, parameters>();
-    test_flystel_prime_field_gadget<ppT, parameters>();
+    test_flystel_E_root_five_gadget<ppT, parameters_debug>();
+    test_flystel_prime_field_gadget<ppT, parameters_debug>();
+    // Use the original parameters for the full permutation
+    using parameters = anemoi_parameters<ppT>;
+    test_anemoi_permutation_round_prime_field_gadget<ppT, 1, parameters>();
+    test_anemoi_permutation_round_prime_field_gadget<ppT, 2, parameters>();
+    test_anemoi_permutation_round_prime_field_gadget<ppT, 3, parameters>();
+    // TODO code for L=4 is still WIP. The test values match, but a
```

Comment:
Seems like this should be tracked down before merging this.

Reply:
Fixed.